tests/phpunit/tests/user/queryCache.php

```diff
@@ -0,0 +1,778 @@
+
+		$post_id = self::factory()->post->create(
+			array(
+				'post_author' => self::$author_ids[2],
+				'post_status' => 'publish',
+				'post_type'   => 'wptests_pt_public',
+			)
+		);
+
+		$q1 = new WP_User_Query(
+			array(
+				'has_published_posts' => true,
+			)
+		);
+
+		$found    = wp_list_pluck( $q1->get_results(), 'ID' );
+		$expected = array( self::$author_ids[2] );
+
+		$this->assertSameSets( $expected, $found, 'Find author in returned values' );
+
+		wp_delete_post( $post_id, true );
+
+		$q2 = new WP_User_Query(
+			array(
+				'has_published_posts' => true,
+			)
+		);
+
+		$found = wp_list_pluck( $q2->get_results(), 'ID' );
+		$this->assertSameSets( array(), $found, 'Not to find author in returned values' );
+	}
+
+	/**
+	 * @ticket 40613
+	 * @covers ::query
+	 */
+	public function test_has_published_posts_delete_post_order() {
+		register_post_type( 'wptests_pt_public', array( 'public' => true ) );
+
+		$user_id = self::factory()->user->create();
+
+		$post_id = self::factory()->post->create(
+			array(
+				'post_author' => $user_id,
+				'post_status' => 'publish',
+				'post_type'   => 'wptests_pt_public',
+			)
+		);
+
+		$q1 = new WP_User_Query(
+			array(
+				'orderby' => 'post_count',
+			)
+		);
+
+		$found1 = wp_list_pluck( $q1->get_results(), 'ID' );
+		$this->assertContains( $user_id, $found1, 'Find author in returned values in first run of WP_User_Query' );
+
+		wp_delete_post( $post_id, true );
+
+		$q2 = new WP_User_Query(
+			array(
+				'orderby' => 'post_count',
+			)
+		);
+
+		$found2 = wp_list_pluck( $q2->get_results(), 'ID' );
+		$this->assertContains( $user_id, $found1, 'Find author in returned values in second run of WP_User_Query' );
+		$this->assertSameSets( $found1, $found2, 'Not same order' );
+	}
+
+	/**
+	 * @ticket 40613
+	 * @covers ::query
+	 */
+	public function test_meta_query_cache_invalidation() {
+		add_user_meta( self::$author_ids[0], 'foo', 'bar' );
+		add_user_meta( self::$author_ids[1], 'foo', 'bar' );
+
+		$q1 = new WP_User_Query(
+			array(
+				'meta_query' => array(
+					array(
+						'key'   => 'foo',
+						'value' => 'bar',
+					),
+				),
+			)
+		);
+
+		$found    = wp_list_pluck( $q1->get_results(), 'ID' );
+		$expected = array( self::$author_ids[0], self::$author_ids[1] );
+
+		$this->assertSameSets( $expected, $found, 'Asset that results contain authors' );
+
+		delete_user_meta( self::$author_ids[1], 'foo' );
+
+		$q2 = new WP_User_Query(
+			array(
+				'meta_query' => array(
+					array(
+						'key'   => 'foo',
+						'value' => 'bar',
+					),
+				),
+			)
+		);
+
+		$found    = wp_list_pluck( $q2->get_results(), 'ID' );
+		$expected = array( self::$author_ids[0] );
+
+		$this->assertSameSets( $expected, $found, 'Asset that results do not contain author without meta' );
+	}
+
+	/**
+	 * @ticket 40613
+	 * @group ms-required
+	 * @covers ::query
+	 */
+	public function test_get_single_capability_multisite_blog_id() {
+		$blog_id = self::factory()->blog->create();
```

Comment:
Sites are created in a few tests, is it possible to set them up as shared fixtures for multisite installs? This will improve the performance of the test suite.

You know way more about MS and testing MS than I, so let me know if it's not possible.

Reply:
A new site is only created in this one test. So no need for shared data here.

Reply:
I'm seeing additional sites created in these methods too:
* `test_query_should_respect_blog_id`
* * `test_has_published_posts_should_respect_blog_id`